// can-ui/panel-container/panel-container.js
import can from 'can/util/library';
import CanMap from 'can/map/';
import CanEvent from 'can/event/';
import Component from 'can/component/';
/**
 * @module panel-container
 */
import template from './panel.stache!';
import './panel.css!';
/**
 * @constructor panel-container.ViewModel ViewModel
 * @parent panel-container
 * @group panel-container.ViewModel.props Properties
 *
 * @description A `<panel-container />` component's ViewModel
 */
export const ViewModel = CanMap.extend({
  /**
   * @prototype
   */
  define: {
    /**
     * The title to use in the tab for this panel
     * @property {String} panel-container.ViewModel.props.title
     * @parent panel-container.ViewModel.props
     */
    title: {
      type: 'string',
      value: 'Panel'
    },
    /**
     * The heading to use in the collapsible header
     * @property {String} panel-container.ViewModel.props.heading
     * @parent panel-container.ViewModel.props
     */
    heading: {
      type: 'string',
      value: 'Panel'
    },
    /**
     * Whether or not this panel is visible by default in the tab container
     * @property {Boolean} panel-container.ViewModel.props.visible
     * @parent panel-container.ViewModel.props
     */
    visible: {
      type: 'boolean',
      value: true
    },
    /**
     * Whether or not this panel is collapsed by default
     * @property {Boolean} panel-container.ViewModel.props.open
     * @parent panel-container.ViewModel.props
     */
    open: {
      type: 'boolean',
      value: true
    },
    /**
     * Whether or not this panel is collapsible
     * @property {Boolean} panel-container.ViewModel.props.collapsible
     * @parent panel-container.ViewModel.props
     */
    collapsible: {
      type: 'boolean',
      value: false
    }
  },
  /**
   * Opens or closes this panel if it is collapsible. Otherwise this will have no effect.
   * @signature
   * @return {can.Map} this view model
   */
  toggle: function() {
    this.attr('open', !this.attr('open'));
    return this;
  },
  /**
   * Hides this panel in a tab-container.
   * @signature
   * @return {can.Map} this view model
   */
  hide: function() {
    this.attr('visible', false);
    this.dispatch('hide', [this]);
    return this;
  },
  /**
   * Shows this panel in a tab-container
   * @signature
   * @return {can.Map} this view model
   */
  show: function() {
    this.attr('visible', true);
    this.dispatch('show', [this]);
    return this;
  }
});
can.extend(ViewModel.prototype, CanEvent);

export default Component.extend({
  tag: 'panel-container',
  template: template,
  viewModel: ViewModel,
  events: {
    inserted: function() {
      if (this.element.parent().viewModel().addPanel) {
        this.element.parent().viewModel().addPanel(this.viewModel);
      }
    },
    removed: function() {
      if (this.element.parent().viewModel().removePanel) {
        this.element.parent().viewModel().removePanel(this.viewModel);
      }
    }
  }
});
